cpp/include/raft/util/arch.cuh

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+namespace raft::arch {
+
+/* raft::arch provides the following facilities:
+ *
+ * - raft::arch::SM_XX : hardcoded compile-time constants for various compute
+ *   architectures. The values raft::arch::SM_min and raft::arch::SM_future
+ *   represent architectures that are always smaller and larger (respectively)
+ *   than any architecture that can be encountered in practice.
+ *
+ * - raft::arch::SM_compute_arch : a compile-time value for the *current*
+ *   compute architecture that a kernel is compiled with. It can only be used
+ *   inside kernels with a template argument.
+ *
+ * - raft::arch::kernel_runtime_arch : a function that computes at *run-time*
+ *   which version of a kernel will launch (i.e., it will return the compute
+ *   architecture of the version of the kernel that will be launched by the
+ *   driver).
+ *
+ * - raft::arch::SM_range : a compile-time value to represent an open interval
+ *   of compute architectures. This can be used to check if the current
+ *   compile-time architecture is in a specified compatibility range.
+ */
+
+// detail::SM_generic is a template to create a generic compile-time SM
+// architecture constant.
+namespace detail {
+template <int n>
+struct SM_generic {
+ public:
+  __host__ __device__ constexpr int value() const { return n; }
+};
+
+// A dummy kernel that is used to determine the runtime architecture.
+__global__ inline void dummy_runtime_kernel() {}
```

Comment:
This needs to be static so we don't run into the issue where multiple consumers of raft build with different `arch` values and we get incorrect kernel selection.

For more info see: https://github.com/NVIDIA/cub/issues/545

Reply:
That's a good point.  It looks like the dummy kernel approach requires making the kernel static to get a reliable solution, at the cost of littering the final binary with many empty kernels. 

In [kernel_runtime_arch](https://github.com/ahendriksen/raft/blob/1a6636f994d3e6e48cd955ecd9ab59c159a54a1c/cpp/include/raft/util/arch.cuh#L114), we are currently taking a pointer to the `dummy_runtime_kernel`. If instead, we took a runtime argument that was a pointer to one of the candidate kernels that is going to be called, would that solve the problem? That is, I would remove the `dummy_runtime_kernel` and the kernel pointer would have to be provided by the user. I think it does solve the linking problem that you described above and it doesn't create spurious kernels, but I want to double check before I change the code.

Reply:
Thanks a lot! I will go for that direction then.

Reply:
I'm a little late to the party, but I came up with an idea for an alternative way of doing this that I like better because it avoids the empty kernel. See https://github.com/NVIDIA/cub/issues/556

Reply:
Thanks for the pointer!  I've been meaning to respond to this for a while, but never found the time to test my assertions. 

We are currently (that is: in the PR that was merged) avoiding the empty kernel by forcing the caller to provide a pointer to one of the kernel versions. We then query the func attributes of that kernel.

The `__CUDA_ARCH_LIST__`  looks like a worthwile approach. However, it may break when kernels are weakly linked (e.g. templated). You describe the issue very well in https://github.com/rapidsai/raft/issues/1722. I had not considered outlawing weak linking completely.. Let's see how that goes!